File: agt/packages.py
"""Combinatorial auctions: bids for *bundles* of items rather than for one item.

A :class:`PackageBid` is an all-or-nothing offer — "I will pay 10 for the northern and
southern licences together, and nothing for either alone". That is what makes this
different from running several single-item auctions side by side: a bidder who needs both
runways, or both halves of a radio band, is worth more with the pair than the two halves
are worth apart, and no sequence of separate auctions lets them say so.

**XOR semantics.** A bidder may submit several package bids, and they are alternatives:
at most one of them can win. The item universe is simply the union of every item anybody
mentioned; there is no separate list of what is for sale.

The module holds two winner-determination algorithms because the honest answer is
expensive. Choosing the best set of bids to accept is NP-hard — it is weighted set
packing — so real combinatorial auctions run something greedy and accept that they left
welfare on the table. :func:`greedy_allocate` is that practical algorithm and
:func:`optimal_allocate` is the exhaustive one, kept next to each other so the cost of
the approximation is a number on the screen rather than a claim in a footnote.
"""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any

from agt.trace import Number

# ponytail: hard caps on the exhaustive search rather than a time budget or a real
# branch-and-bound solver, because a cap is one comparison and a time budget makes the
# answer depend on how busy the machine is — and a teaching example that silently
# returned a different allocation on a slow day would be worse than one that refuses.
# Measured worst case at this bound is ~5ms (20 near-equal singleton bids from 20
# bidders over 12 items, ~10.9k search nodes), so the whole search fits inside one HTTP
# request many times over. Ceiling: a real spectrum auction has hundreds of items.
# Upgrade: swap the body of `optimal_allocate` for an LP relaxation with branch-and-bound.
MAX_ITEMS = 12
MAX_BIDS = 20


@dataclass(frozen=True)
class PackageBid:
    """One all-or-nothing offer: ``bid`` for exactly ``items``, worth ``value`` to win.

    ``value`` is private and only the engine sees it, exactly as in :class:`~agt.trace.Bidder`;
    ``bid`` is what the mechanism is allowed to look at.
    """

    bidder: str
    items: tuple[str, ...]
    value: Number
    bid: Number

    def to_dict(self) -> dict[str, Any]:
        return {
            "bidder": self.bidder,
            "items": list(self.items),
            "value": self.value,
            "bid": self.bid,
        }


Bids = Sequence[PackageBid]
Allocation = tuple[PackageBid, ...]


def item_universe(bids: Bids) -> tuple[str, ...]:
    """Everything anybody bid on, sorted, each item once."""
    return tuple(sorted({item for b in bids for item in b.items}))


def total_bid(accepted: Bids) -> Number:
    """What an allocation is worth to the auctioneer, in submitted bids."""
    return sum(b.bid for b in accepted)


def _ranked(bids: Bids) -> list[PackageBid]:
    """Highest bid first. ``sorted`` is stable, so ties keep the listed order — the same
    tie rule the single-item mechanisms use, and the reason both solvers are reproducible."""
    for b in bids:
        if not b.items:
            raise ValueError(
                f"{b.bidder!r} submitted a package of no items; a package bid must name "
                "at least one item, or it would conflict with nothing and win for free"
            )
    return sorted(bids, key=lambda b: -b.bid)


def _masks(ranked: list[PackageBid]) -> list[tuple[int, int]]:
    """One (items, bidder) bit pair per bid, so a conflict is a single ``&``."""
    item_bit = {name: 1 << i for i, name in enumerate(item_universe(ranked))}
    owner_bit = {
        name: 1 << i for i, name in enumerate(dict.fromkeys(b.bidder for b in ranked))
    }
    return [
        (sum(item_bit[i] for i in set(b.items)), owner_bit[b.bidder]) for b in ranked
    ]


def _pack(masks: list[tuple[int, int]], order: range) -> tuple[int, ...]:
    """The greedy rule itself: accept in ``order``, skip anything that conflicts.

    Two things count as a conflict — the bundle shares an item with something already
    accepted, or its bidder has already won a different bundle (XOR).
    """
    items = owners = 0
    chosen: list[int] = []
    for j in order:
        bundle, owner = masks[j]
        if items & bundle or owners & owner:
            continue
        chosen.append(j)
        items |= bundle
        owners |= owner
    return tuple(chosen)


def greedy_allocate(bids: Bids) -> Allocation:
    """Accept bids in descending order, skipping every one that conflicts. Linearithmic.

    This is what a practical auction runs, and it is not optimal — a single large bid
    taken early can block a set of smaller ones worth more together.
    """
    ranked = _ranked(bids)
    return tuple(ranked[j] for j in _pack(_masks(ranked), range(len(ranked))))


def optimal_allocate(bids: Bids) -> Allocation:
    """The allocation with the highest total bid. Exponential, hence the size guard.

    Depth-first over the same descending order greedy walks, carrying the used items and
    the bidders who have already won as bitmasks, and cutting any branch whose remaining
    bids could not add up past the best allocation found so far.

    The search *starts* from greedy's answer and only ever replaces it on a strict
    improvement. That is worth two things: greedy is a real lower bound, so seeding with
    it prunes far more of the tree than starting from nothing, and it settles the
    tie-break — among allocations worth exactly the same, the one greedy would have
    picked is the one returned, so the answer is reproducible.
    """
    ranked = _ranked(bids)
    _check_size(ranked)
    masks = _masks(ranked)

    n = len(ranked)
    remaining = [0] * (n + 1)  # the most the bids from index i onwards could ever add
    for i in range(n - 1, -1, -1):
        remaining[i] = remaining[i + 1] + ranked[i].bid

    seed = _pack(masks, range(n))
    best: tuple[Number, tuple[int, ...]] = (sum(ranked[j].bid for j in seed), seed)

    def search(
        start: int, items: int, owners: int, total: Number, chosen: tuple[int, ...]
    ) -> None:
        nonlocal best
        if total > best[0]:
            best = (total, chosen)
        for j in range(start, n):
            if total + remaining[j] <= best[0]:
                return  # nothing left in this branch can beat what we already have
            bundle, owner = masks[j]
            if items & bundle or owners & owner:
                continue
            search(j + 1, items | bundle, owners | owner, total + ranked[j].bid, chosen + (j,))

    search(0, 0, 0, 0, ())
    return tuple(ranked[j] for j in best[1])


def _check_size(bids: Bids) -> None:
    """Refuse inputs the exhaustive search cannot be trusted to finish quickly.

    This guard is the honest face of NP-hardness: rather than hang inside a web request,
    the auction says out loud how big a problem it is willing to solve exactly.
    """
    items = item_universe(bids)
    if len(items) > MAX_ITEMS or len(bids) > MAX_BIDS:
        raise ValueError(
            "winner determination is NP-hard, so the exhaustive search is capped at "
            f"{MAX_ITEMS} items and {MAX_BIDS} bids; this input has items="
            f"{len(items)}, bids={len(bids)}. Greedy allocation has no such limit."
        )

File: web/packages.js
'use strict';

/* ===========================================================================
   Combinatorial auctions — bidding on bundles instead of on one thing.

   THE ONE RULE from app.js still holds: no mechanism name appears below. The
   setup panel switches because the registry says the selected mechanism has
   `input_kind: "package"`, and the allocation view draws because the trace's
   allocation hands out lists of items instead of slot numbers. Shape and
   declaration decide; names never do.

   Two seams into app.js: `runAuction()` hands the whole submit path over here
   for package mechanisms (they post `packages`, not `bidders`), and the
   mechanism <select> tells us to swap the panel. Everything else — `api`,
   `render`, `traceScale`, `showError`, `app` — is used exactly as app.js
   exposes it, and nothing in app.js's state is written from here except
   `app.trace` / `app.index` via its own `render`.
   =========================================================================== */

// The instance that makes the point on first load: A wants both licences and bids
// 10; B and C each want one and bid 6. Greedy takes A's single big bid and is then
// blocked, scoring 10; the optimum splits the pair for 12. The gap is the reason
// this whole sub-phase exists, so the page opens already showing it.
const DEFAULT_PACKAGES = [
  { bidder: 'A', items: 'north, south', value: 10, bid: 10 },
  { bidder: 'B', items: 'north', value: 6, bid: 6 },
  { bidder: 'C', items: 'south', value: 6, bid: 6 },
];

// Mirrors the engine's own limits (agt/winner_determination.py). Checked here so a
// too-large problem is refused next to the form rather than after a round trip; the
// server checks them again, because a browser is not a trust boundary.
const MAX_BIDS = 20;
const MAX_ITEMS = 12;

/** Items are typed as free text; the universe is whatever was typed. Commas or
 *  spaces both separate, because people reach for either. */
function parseItems(raw) {
  return raw.split(/[,\s]+/).map((s) => s.trim()).filter(Boolean);
}

const currentSpec = () => (app.registry || {})[$('mechanism').value] || {};

// ============================================================ the bid table =

function addPackageRow(seed) {
  const tbody = $('package-rows');
  const tr = h('tr');

  const cell = (field, type, value, extra) => {
    const td = h('td', extra || null);
    const input = h('input');
    input.type = type;
    input.value = value;
    input.dataset.field = field;
    input.setAttribute('aria-label', field);
    if (type === 'number') { input.step = 'any'; input.min = '0'; }
    if (field === 'bidder') input.addEventListener('input', markAlternatives);
    td.append(input);
    tr.append(td);
    return input;
  };

  cell('bidder', 'text', seed.bidder);
  cell('items', 'text', seed.items, 'items-cell');
  cell('value', 'number', seed.value);
  cell('bid', 'number', seed.bid);

  const td = h('td');
  const remove = h('button', 'remove', '✕');
  remove.type = 'button';
  remove.title = `Remove ${seed.bidder}'s bid for ${seed.items}`;
  remove.addEventListener('click', () => {
    if (tbody.children.length <= 1) return;   // an auction needs at least one bid
    tr.remove();
    markAlternatives();
  });
  td.append(remove);
  tr.append(td);

  tbody.append(tr);
  markAlternatives();
}

/** "I can win at most one of these" is the part of XOR bidding people get wrong,
 *  so rows that share a bidder are marked as alternatives in the table itself
 *  rather than explained once in prose above it. */
function markAlternatives() {
  const rows = [...$('package-rows').children];
  const counts = new Map();
  for (const tr of rows) {
    const name = (tr.querySelector('[data-field="bidder"]').value || '').trim();
    counts.set(name, (counts.get(name) || 0) + 1);
  }

  let grouped = 0;
  for (const tr of rows) {
    const name = (tr.querySelector('[data-field="bidder"]').value || '').trim();
    const isAlt = name !== '' && counts.get(name) > 1;
    tr.classList.toggle('xor-alt', isAlt);
    const cell = tr.querySelector('[data-field="bidder"]').parentElement;
    let tag = cell.querySelector('.xor-tag');
    if (isAlt && !tag) {
      tag = h('span', 'xor-tag', 'either/or');
      tag.title = `${name} submitted more than one bundle and can win at most one`;
      cell.append(tag);
    } else if (!isAlt && tag) {
      tag.remove();
    }
    if (isAlt) grouped += 1;
  }

  const note = $('package-note');
  note.textContent = grouped > 0
    ? `${grouped} rows are marked either/or: those bidders submitted alternative bundles and can win at most one each.`
    : 'Give two rows the same bidder name to offer alternative bundles — that bidder can then win at most one of them.';
}

function readPackages() {
  return [...$('package-rows').children].map((tr) => {
    const get = (field) => tr.querySelector(`[data-field="${field}"]`).value;
    return {
      bidder: get('bidder').trim(),
      items: parseItems(get('items')),
      value: Number(get('value')),
      bid: Number(get('bid')),
    };
  });
}

function validatePackages(bids) {
  if (bids.length === 0) return 'Add at least one package bid.';
  if (bids.length > MAX_BIDS) {
    return `${bids.length} package bids is more than the exhaustive search will take on. The limit is ${MAX_BIDS}.`;
  }

  const universe = new Set();
  for (const b of bids) {
    if (!b.bidder) return 'Every package bid needs a bidder name.';
    if (b.items.length === 0) return `${b.bidder} needs at least one item — a bundle of nothing would conflict with nothing and win for free.`;
    if (new Set(b.items).size !== b.items.length) {
      return `${b.bidder} listed the same item twice in one bundle.`;
    }
    if (!isNum(b.value) || b.value < 0) return `${b.bidder} needs a value of 0 or more.`;
    if (!isNum(b.bid) || b.bid < 0) return `${b.bidder} needs a bid of 0 or more.`;
    for (const item of b.items) universe.add(item);
  }

  if (universe.size > MAX_ITEMS) {
    return `${universe.size} distinct items is more than the exhaustive search will take on. The limit is ${MAX_ITEMS}.`;
  }
  return null;
}

// ================================================================ the panel =

/** Swap the setup panel between the scalar bidder table and the package table,
 *  and hide the series controls — a strategy produces one number, which means
 *  nothing over bundles, and POST /run_series refuses these mechanisms outright. */
function applyMode() {
  const isPackage = currentSpec().input_kind === 'package';

  $('single-setup').hidden = isPackage;
  $('package-setup').hidden = !isPackage;

  for (const selector of ['.series-run', '#strategy-notes']) {
    const node = document.querySelector(selector);
    if (node) node.hidden = isPackage;
  }
  if (isPackage) {
    $('timeline-card').hidden = true;
    $('series-card').hidden = true;
    $('decision-card').hidden = true;
  }
}

// ============================================================== the results =

/** The bundles that were won, in the order the engine ranked them, or null when
 *  this trace allocated slot numbers (the ladder's job) or nothing at all. */
function bundleRows(trace) {
  const result = (trace && trace.result) || {};
  const allocation = result.allocation;
  if (!allocation || typeof allocation !== 'object' || Array.isArray(allocation)) return null;

  const entries = Object.entries(allocation);
  if (!entries.length || !entries.every(([, items]) => Array.isArray(items))) return null;

  const payments = result.payments || {};
  const utilities = result.utilities || {};
  const gains = result.gains || {};
  return entries.map(([id, items]) => ({
    id,
    items,
    pays: isNum(payments[id]) ? payments[id] : 0,
    gain: isNum(gains[id]) ? gains[id] : 0,
    utility: isNum(utilities[id]) ? utilities[id] : 0,
  }));
}

/** Two bars on one welfare scale: what greedy found, and what was there to find.
 *  Not two charts and not two axes — the whole point is that they are the same
 *  measure, so the shortfall is a length you can see rather than a subtraction
 *  you have to do. Colour reuses the page's validated ordinal pair (the strong
 *  step for what was achieved, the light step for what was available) instead of
 *  opening a second colour language; identity comes from each row's own label,
 *  so there is no legend and no colour-only encoding. */
function renderGap(result) {
  const greedy = result.greedy_welfare;
  const optimal = result.optimal_welfare;
  const panel = $('gap-panel');

  if (!isNum(greedy) || !isNum(optimal)) {
    panel.replaceChildren();
    panel.hidden = true;
    $('gap-note').hidden = true;
    return;
  }
  panel.hidden = false;

  const gap = isNum(result.welfare_gap) ? result.welfare_gap : optimal - greedy;
  const max = Math.max(greedy, optimal, 1);

  const hero = h('div', 'gap-hero');
  hero.append(h('span', 'gap-figure', gap > 0 ? `−${fmt(gap)}` : '0'));
  hero.append(h('span', 'gap-caption', gap > 0
    ? 'welfare the greedy rule left on the table'
    : 'greedy found the best allocation here'));
  panel.replaceChildren(hero);

  const bars = h('div', 'gap-bars');
  const row = (label, amount, cls) => {
    const line = h('div', 'gap-row');
    line.append(h('span', 'gap-label', label));

    const plot = h('div', 'gap-plot');
    const bar = h('span', `gap-bar ${cls}`);
    bar.style.width = `${(amount / max) * 100}%`;
    bar.setAttribute('aria-hidden', 'true');   // the value is the cell right beside it
    plot.append(bar);
    line.append(plot);

    line.append(h('span', 'gap-value', fmt(amount)));
    bars.append(line);
  };
  row('Greedy took', greedy, 'is-greedy');
  row('Best possible', optimal, 'is-optimal');
  panel.append(bars);

  // Two numbers on one screen that are measured differently will read as a
  // contradiction unless the difference is stated. `welfare` scores what the
  // items were worth to whoever got them; these two total the submitted bids,
  // which is what the solver actually maximises. They agree only under truthful
  // bidding — which is exactly what one of these two mechanisms does not get.
  const note = $('gap-note');
  note.hidden = false;
  note.textContent = 'These two totals are in submitted bids — the quantity the solver maximises. ' +
    'The Outcome card scores welfare in private values instead, so the two agree only when bidding is truthful.';
}

function renderBundles(trace, rows) {
  const result = trace.result || {};

  const table = h('table', 'data-table bundle-table');
  const headRow = h('tr');
  for (const label of ['Bidder', 'Bundle won', 'Pays', 'Utility']) {
    const th = h('th', null, label);
    th.scope = 'col';
    headRow.append(th);
  }
  const thead = h('thead');
  thead.append(headRow);
  table.append(thead);

  const tbody = h('tbody');
  for (const row of rows) {
    const tr = h('tr');
    tr.append(h('td', null, star(row.id, row.id === result.winner)));
    tr.append(h('td', 'items-cell', row.items.join(' + ')));
    tr.append(h('td', null, fmt(row.pays)));
    tr.append(h('td', isNum(row.utility) && row.utility < 0 ? 'neg' : null, fmt(row.utility)));
    tbody.append(tr);
  }
  table.append(tbody);
  $('bundle-table').replaceChildren(table);

  // What nobody won is as much the outcome as what somebody did — an item left
  // unsold is welfare that simply did not happen.
  const universe = itemUniverse(trace);
  const taken = new Set(rows.flatMap((r) => r.items));
  const unsold = universe.filter((item) => !taken.has(item));
  $('unsold-note').textContent = unsold.length
    ? `Unsold: ${unsold.join(', ')} — nobody who bid for ${unsold.length === 1 ? 'it' : 'them'} could be fitted into the allocation.`
    : 'Every item found a buyer.';
}

/** The engine publishes the item universe on the step that establishes it; fall
 *  back to the bids themselves so the unsold line still works if it ever stops. */
function itemUniverse(trace) {
  for (const step of trace.steps) {
    if (Array.isArray(step.state && step.state.items)) return step.state.items;
  }
  const seen = new Set();
  for (const bid of trace.bidders || []) for (const item of bid.items || []) seen.add(item);
  return [...seen];
}

// =================================================================== submit =

async function runPackageAuction() {
  const bids = readPackages();
  const problem = validatePackages(bids);
  if (problem) { showError(problem); return; }

  const button = $('run');
  button.disabled = true;
  button.textContent = 'Running…';
  try {
    const trace = await api('/run', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ mechanism: $('mechanism').value, packages: bids, params: readParams() }),
    });
    if (!trace || !Array.isArray(trace.steps) || trace.steps.length === 0) {
      throw new Error('The server returned a trace with no steps.');
    }
    app.trace = trace;
    app.scale = traceScale(trace);
    render(0);
  } catch (err) {
    showError(err.message);
  } finally {
    button.disabled = false;
    button.textContent = 'Run auction';
  }
}

// ===================================================================== boot =

window.packagesExt = {
  active() {
    return currentSpec().input_kind === 'package';
  },

  boot() {
    for (const seed of DEFAULT_PACKAGES) addPackageRow(seed);
    $('add-package').addEventListener('click', () => {
      addPackageRow({ bidder: 'A', items: '', value: 10, bid: 10 });
    });
    applyMode();
  },

  onMechanismChange() {
    applyMode();
  },

  runAuction: runPackageAuction,

  /** Called at the end of every app.js render, for every trace. */
  onRender(trace, index) {
    const card = $('packages-card');
    if (!card) return;

    const rows = bundleRows(trace);
    // Utilities settle only at the end, so this appears exactly when the Outcome
    // card does — one rule for "the auction is over", not two. Same choice the
    // slot ladder makes, for the same reason.
    // ponytail: the allocation does not fill in bundle by bundle as the per-winner
    // price steps run. Ceiling: the arithmetic for one bundle is legible in the
    // step panel, not here. Upgrade: draw rows from step state, utility last.
    const finished = index === trace.steps.length - 1;

    card.hidden = !rows || !finished;
    if (card.hidden) return;

    renderGap(trace.result || {});
    renderBundles(trace, rows);
  },
};
